File: phases/18-language-paradigms-and-type-theory/15-phase-capstone-a-bidirectional-type-checker-for-stlc-extensions/code/Main.hs
module Main where

import qualified Data.Map as M

data Ty = TBool | TArr Ty Ty deriving (Eq, Show)
data Tm = TV String | TLam String Tm | TApp Tm Tm | TAnn Tm Ty | TTrue deriving (Eq, Show)
type Ctx = M.Map String Ty

infer :: Ctx -> Tm -> Either String Ty
infer g (TV x) = maybe (Left $ "unbound: " ++ x) Right (M.lookup x g)
infer _ TTrue = Right TBool
infer g (TAnn t ty) = do check g t ty; Right ty
infer g (TApp f a) = do
  tf <- infer g f
  case tf of
    TArr i o -> do check g a i; Right o
    _ -> Left "apply non-function"
infer _ (TLam _ _) = Left "lambda needs expected type"

check :: Ctx -> Tm -> Ty -> Either String ()
check g (TLam x body) (TArr i o) = check (M.insert x i g) body o
check g t ty = do
  it <- infer g t
  if it == ty then Right () else Left $ "expected " ++ show ty ++ ", got " ++ show it

main :: IO ()
main = do
  let idAnn = TAnn (TLam "x" (TV "x")) (TArr TBool TBool)
  print (infer M.empty idAnn)
  print (infer M.empty (TApp idAnn TTrue))

File: phases/18-language-paradigms-and-type-theory/15-phase-capstone-a-bidirectional-type-checker-for-stlc-extensions/code/main.rs
use std::collections::HashMap;

#[derive(Clone, Debug, PartialEq, Eq)]
enum Ty { Bool, Arr(Box<Ty>, Box<Ty>) }

#[derive(Clone, Debug)]
enum Tm {
    V(&'static str),
    Lam(&'static str, Box<Tm>),
    App(Box<Tm>, Box<Tm>),
    Ann(Box<Tm>, Ty),
    True,
}

type Ctx = HashMap<&'static str, Ty>;

fn infer(ctx: &Ctx, tm: &Tm) -> Result<Ty, String> {
    match tm {
        Tm::V(x) => ctx.get(x).cloned().ok_or_else(|| format!("unbound: {}", x)),
        Tm::True => Ok(Ty::Bool),
        Tm::Ann(t, ty) => { check(ctx, t, ty)?; Ok(ty.clone()) }
        Tm::App(f, a) => match infer(ctx, f)? {
            Ty::Arr(i, o) => { check(ctx, a, &i)?; Ok(*o) }
            _ => Err("apply non-function".into()),
        },
        Tm::Lam(_, _) => Err("lambda needs expected type".into()),
    }
}

fn check(ctx: &Ctx, tm: &Tm, ty: &Ty) -> Result<(), String> {
    match (tm, ty) {
        (Tm::Lam(x, body), Ty::Arr(i, o)) => {
            let mut c = ctx.clone(); c.insert(x, (*i.clone()).clone());
            check(&c, body, o)
        }
        _ => {
            let it = infer(ctx, tm)?;
            if &it == ty { Ok(()) } else { Err(format!("expected {:?}, got {:?}", ty, it)) }
        }
    }
}

fn main() {
    let id = Tm::Ann(Box::new(Tm::Lam("x", Box::new(Tm::V("x")))), Ty::Arr(Box::new(Ty::Bool), Box::new(Ty::Bool)));
    let app = Tm::App(Box::new(id), Box::new(Tm::True));
    println!("{:?}", infer(&Ctx::new(), &app));
}
